File: go/day16.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

func two() {
	fmt.Println("hello")
}

type valve struct {
	name string
	rate int
	next []string
}

func parse(path string) []valve {
	path, err := filepath.Abs(path)
	if err != nil {
		fmt.Println(err)
		return []valve{}
	}
	fmt.Println(path)

	f, err := os.OpenFile(path, os.O_RDONLY, os.ModePerm)
	if err != nil {
		fmt.Println(err)
		return []valve{}
	}
	defer f.Close()

	re := regexp.MustCompile(`(?m)Valve (\w+).+rate=(\d+).+valves? ([A-Z, ]+)`)
	sc := bufio.NewScanner(f)

	valves := make([]valve, 0)

	var line []byte
	var matches [][]byte
	var v valve
	for sc.Scan() {
		line = sc.Bytes()
		matches = re.FindSubmatch(line)

		rate, err := strconv.Atoi(string(matches[2]))
		if err != nil {
			fmt.Println(err)
			return []valve{}
		}

		v = valve{
			name: string(matches[1]),
			rate: rate,
			next: strings.Split(string(matches[3]), ", "),
		}

		valves = append(valves, v)
	}

	return valves
}

func calcPath(valveMap map[string]valve, from, to string) int {
	if from == to {
		return 0
	}

	queue := []string{from}
	distance := make(map[string]int)
	for k := range valveMap {
		distance[k] = -1
	}
	distance[from] = 0

	var current string
	for {
		current = queue[0]
		queue = queue[1:]

		for _, next := range valveMap[current].next {
			if distance[next] > -1 {
				continue
			}

			distance[next] = distance[current] + 1
			queue = append(queue, next)

			if next == to {
				return distance[to]
			}
		}
	}
}

type move struct {
	who     string
	current string
	time    int
}

func exclude(ss []string, excl string) []string {
	for i, s := range ss {
		if s == excl {
			newS := make([]string, len(ss[:i]))
			copy(newS, ss[:i])
			return append(newS, ss[i+1:]...)
		}
	}

	return ss
}

func iter(valveMap map[string]valve, opened []string, path []move) [][]move {
	if len(opened) == 0 {
		return [][]move{path}
	}

	var lstH, lstE, lst move
	for _, m := range path {
		if m.who == "H" && (lstH.who == "" || lstH.time < m.time) {
			lstH = m
		}
		if m.who == "E" && (lstE.who == "" || lstE.time < m.time) {
			lstE = m
		}
	}

	if lstH.time < lstE.time {
		lst = lstH
	} else {
		lst = lstE
	}

	res := make([][]move, 0)
	var cost int
	for _, o := range opened {
		cost = calcPath(valveMap, lst.current, o) + 1

		if lst.time+cost > 26 {
			continue
		}

		newPath := make([]move, len(path))
		copy(newPath, path)

		res = append(
			res,
			iter(
				valveMap,
				exclude(opened, o),
				append(newPath, move{who: lst.who, current: o, time: lst.time + cost}),
			)...,
		)
	}

	return res
}

func main() {
	valves := parse("./resources/2022/day16.csv")
	valveMap := make(map[string]valve)

	opened := make([]string, 0)
	for _, v := range valves {
		valveMap[v.name] = v
		if v.rate > 0 {
			opened = append(opened, v.name)
		}
	}

	// fmt.Printf("%+v\n", valves)

	res := iter(valveMap, opened, []move{
		{who: "H", current: "AA", time: 0},
		{who: "E", current: "AA", time: 0},
	})

	var max, maxI, sum int
	for i, path := range res {
		sum = 0
		for _, v := range path {
			sum += valveMap[v.current].rate * (26 - v.time)
		}

		if sum > max {
			max = sum
			maxI = i
		}
	}

	fmt.Println(max)
	// fmt.Printf("%+v\n", res[maxI])
	fmt.Println(res[maxI])
}

File: src/2022/day16.js
import { fileLines, fileLinesBy, RESOURCES } from '../index.js';
import createGraph from 'ngraph.graph'
import path from 'ngraph.path'

const FILEPATH = '2022/day16.csv'
const FILEPATH_TEST = '2022/day16.test.csv'

function parse(line) {
  const match = line.match(/Valve (?<name>\w+).+rate=(?<rate>\d+).+valves? (?<next>[A-Z, ]+)/)
  return {
    name: match.groups.name,
    rate: parseInt(match.groups.rate),
    next: match.groups.next.split(', '),
  }
}

function diff(a, b) {
  return a.filter(i => !b.includes(i))
}

function calcPath(valveMap, a, b) {
  if (a === b) {
    return 0
  }

  let distance = Object.fromEntries(Object.keys(valveMap).map(k => [k, -1]))
  distance[a] = 0
  let queue = [a]

  while (true) {
    const current = queue.shift()
    const notVisited = valveMap[current].next.filter(n => distance[n] === -1)

    notVisited.forEach(v => {
      distance[v] = distance[current] + 1
      queue.push(v)
    })

    if (notVisited.includes(b)) {
      return distance[b]
    }
  }
}

function memo(fn, ...def) {
  const mem = {}

  return (...args) => {
    const argsHash = JSON.stringify([...def, ...args])

    if (!(argsHash in mem)) {
      mem[argsHash] = fn(...def, ...args)
    }

    return mem[argsHash]
  }
}

async function one() {
  const lines = await fileLines(FILEPATH_TEST)

  const valves = lines.map(parse)
  const valveMap = Object.fromEntries(valves.map(({name, ...valve}) => [name, valve]))

  const memoPath = memo(calcPath, valveMap)

  function iter(current, opened, dis, acc) {
    const possible = opened.map(n => [n, memoPath(current, n)]).filter(([_, cost]) => cost < dis)

    if (possible.length === 0) {
      return [...acc, current].join('|')
    }

    const res = possible.flatMap(([next, cost]) => iter(
      next,
      diff(opened, [next]),
      dis - cost - 1,
      [...acc, current]
    ))

    return res
  }

  const opened = valves.filter(({rate}) => rate > 0).map(({name}) => name)
  const paths = iter('AA', opened, 30, [])

  let max = 0
  let maxPath = ""
  paths.forEach(path => {
    const vs = path.split('|')

    let acc = 0
    let rate = 0
    let time = 30
    for (let i = 0; i < vs.length - 1; i++) {
      const [a, b] = [vs[i], vs[i+1]]
      const cost = memoPath(a, b)
      acc += rate * (cost + 1)
      rate += valveMap[b].rate
      time -= cost + 1
    }

    acc += rate * time

    if (acc > max) {
      max = acc
      maxPath = path
    }
  })

  console.log(max)
  console.log(maxPath)
}

function lastBy(arr, cb) {
  const keys = []
  const res = []

  for (let i = arr.length-1; i>=0; i--) {
    const key = cb(arr[i])
    if (keys.includes(key)) {
      continue
    }
    keys.push(key)
    res[keys.indexOf(key)] = arr[i]
  }

  return res.reverse();
}

async function two() {
  const lines = await fileLines(FILEPATH_TEST)

  const valves = lines.map(parse)
  const valveMap = Object.fromEntries(valves.map(({name, ...valve}) => [name, valve]))

  const memoPath = memo(calcPath, valveMap)

  let count = 0

  function iter(opened, acc) {
    const lastes = lastBy(acc, i => i.who).sort((a, b) => a.time - b.time)
    const last = lastes[0]

    if (acc.some(i => i.time > 26)) {
      return [acc.filter(i => i.time <= 26)]
    }

    if (opened.length === 0) {
      return [acc]
    }

    const res = opened.map(next => {
      const cost = memoPath(last.current, next) + 1

      return iter(
        diff(opened, [next]),
        [...acc, {who: last.who, current: next, time: last.time+cost}]
      )
    })

    return res.flat(1)
  }

  const opened = valves.filter(({rate}) => rate > 0).map(({name}) => name)
  const paths = iter(opened, [{who: 'E', current: 'AA', time: 0}, {who: 'H', current: 'AA', time: 0}])

  console.log(paths.length)
  let max = 0
  let maxPath
  paths.map(path => {
    const rates = path.map(({current, time}) => valveMap[current].rate * (26-time)).reduce((acc, i) => acc + i, 0)
    if (rates > max) {
      max = rates
      maxPath = path
    }
  })

  console.log(max)
  console.log(maxPath)
  return
}

(async function () {
  // await one()
  await two()
})()
